src/shared/nscd-flush.c

```diff
@@ -0,0 +1,127 @@
+/* SPDX-License-Identifier: LGPL-2.1+ */
+#include <sys/poll.h>
+
+#include "alloc-util.h"
+#include "fd-util.h"
+#include "io-util.h"
+#include "nscd-flush.h"
+#include "socket-util.h"
+#include "strv.h"
+#include "time-util.h"
+#include "unaligned.h"
+
+#define NSCD_FLUSH_CACHE_TIMEOUT_USEC (5*USEC_PER_SEC)
+
+struct nscdInvalidateRequest {
+        int32_t version;
+        int32_t type; /* in glibc this is an enum. We don't replicate this here 1:1. Also, wtf, how unportable is that even? */
+        int32_t key_len;
+        char dbname[];
+};
+
+static const union sockaddr_union nscd_sa = {
+        .un.sun_family = AF_UNIX,
+        .un.sun_path = "/run/nscd/socket",
+};
+
+static int nscd_flush_cache_one(const char *database, usec_t end) {
+        size_t req_size, has_written = 0, has_read = 0, l;
+        struct nscdInvalidateRequest *req;
+        _cleanup_close_ int fd = -1;
+        int32_t resp;
+        int events;
+
+        l = strlen(database);
+        req_size = offsetof(struct nscdInvalidateRequest, dbname) + l + 1;
+
+        req = alloca(req_size);
+        *req = (struct nscdInvalidateRequest) {
+                .version = 2,
+                .type = 10,
+                .key_len = l + 1,
+        };
+
+        strcpy(req->dbname, database);
+
+        fd = socket(AF_UNIX, SOCK_STREAM|SOCK_CLOEXEC|SOCK_NONBLOCK, 0);
+        if (fd < 0)
+                return log_debug_errno(errno, "Failed to allocate nscd socket: %m");
+
+        if (connect(fd, &nscd_sa.sa, SOCKADDR_UN_LEN(nscd_sa.un)) < 0 && errno != EINPROGRESS)
+                return log_debug_errno(errno, "Failed to connect to nscd socket: %m");
+
+        events = POLLIN|POLLOUT; /* Let's assume initially that we can write and read to the fd, to suppress a poll()
+                                  * invocation */
+        for (;;) {
+                usec_t p;
+
+                if (events & POLLOUT) {
+                        ssize_t m;
+
+                        assert(has_written < req_size);
+
+                        m = send(fd, (uint8_t*) req + has_written, req_size - has_written, MSG_NOSIGNAL);
+                        if (m < 0) {
+                                if (errno != EAGAIN)
```

Comment:
What happens if the connect() returned EINPROGRESS, nscd's behind on accept()s exceeding the backlog, and the socket isn't connected yet?

Won't send() return -1 && errno == ENOTCONN?  My sockets API memory is rusty...

Reply:
I did a quick test, and you're right the current AF_UNIX implementation never returns EINPROGRESS.

Instead, it either immediately connects if backlog hasn't been exceeded, or it returns EAGAIN without actually kicking off a non-blocking connect.

So I think there's still a robustness problem with this PR as-is.  If your connect returns EAGAIN, the flush will simply fail to occur.

Preparing for a future kernel where EINPROGRESS is possible on AF_UNIX non-blocking connects probably isn't a bad idea either.

Reply:
> So I think there's still a robustness problem with this PR as-is. If your connect returns EAGAIN, the flush will simply fail. The whole "nscd behind on accept()s exceeding the backlog" problem is moved up to the connect call instead.

Hmm, so the whole function is written very defensively anyway: it tries to flush the cache on a best-effort basis: if we can't flush the the databases we give up quickly (after a 5s timeout already) and besides debug logging we don't do anything anymore. And I think this is the right approach, since we need to be careful with deadlocks here, after all we are regularly both consumers and providers of nss, and even if we are super careful we might deadlock somewhere.

And I think a backlog overload is an exceptional condition that right-fully can be considered equivalent to a timeout being reached: it just means that things are overloaded in some way...

BTW, I verified that on on AF_INET if socket() returns EINPROGRESS and you immediately issue send() on it, then you'll get EAGAIN. Hence the current code already handled this pretty OK, as in that case it would enter poll() and wait.

Reply:
Ok, and nscd looks to set the backlog pretty high so I guess it's an unlikely enough failure.